lib/semantics/check-do.cc

```diff
@@ -660,116 +658,115 @@ static parser::CharBlock GetNodePosition(const ConstructNode &construct) {
-    } else {
-      return false;  // Unlabeled statements match no non-DO constructs
-    }
-  } else if (!constructName) {
-    return false;  // name on CYCLE/EXIT, but not on the construct
-  } else if (constructName->source == stmtName->source) {
-    if (isExitStmt(stmtType)) {
-      return true;  // EXIT name matches any construct name
-    } else {
-      if (inDoConstruct) {
-        return true;  // CYCLE name matches only DO construct name
-      }
-    }
+    return inDoConstruct;  // Unlabeled statements match all DO constructs
+  } else if (constructName && constructName->source == stmtName->source) {
+    return isExit || inDoConstruct;
+  } else {
+    return false;
   }
-  return false;
 }
 
 // C1167 Can't EXIT from a DO CONCURRENT
 void DoChecker::CheckDoConcurrentExit(
-    const char *stmtType, const ConstructNode &construct) const {
-  if (isExitStmt(stmtType) && ConstructIsDoConcurrent(construct)) {
+    bool isExit, const ConstructNode &construct) const {
+  if (isExit && ConstructIsDoConcurrent(construct)) {
     SayBadLeave("EXIT", "DO CONCURRENT", construct);
   }
 }
 
-// Check nesting violations for a CYCLE or EXIT statement Loop up the nesting
+// Check nesting violations for a CYCLE or EXIT statement.  Loop up the nesting
 // levels looking for a construct that matches the CYCLE or EXIT statment.  At
 // every construct, check for a violation.  If we find a match without finding
 // a violation, the check is complete.
-void DoChecker::CheckNesting(const char *stmtType, NamePtr stmtName) const {
+void DoChecker::CheckNesting(bool isExit, const parser::Name *stmtName) const {
   const ConstructStack &stack{context_.constructStack()};
-  for (ConstructStack::const_reverse_iterator riter = stack.crbegin();
-       riter != stack.crend(); ++riter) {
-    const ConstructNode &construct{*riter};
-    NamePtr constructName{MaybeGetNodeName(construct)};
-    if (StmtMatchesConstruct(stmtName, stmtType, constructName, construct)) {
-      CheckDoConcurrentExit(stmtType, construct);
+  for (auto rIter{stack.crbegin()}; rIter != stack.crend(); ++rIter) {
+    const ConstructNode &construct{*rIter};
+    const parser::Name *constructName{MaybeGetNodeName(construct)};
+    if (StmtMatchesConstruct(stmtName, isExit, constructName, construct)) {
+      CheckDoConcurrentExit(isExit, construct);
       return;  // We got a match, so we're finished checking
     }
-    CheckForBadLeave(stmtType, construct);
+    CheckForBadLeave(GetStmtTypeName(isExit), construct);
   }
 
   // We haven't found a match in the enclosing constructs
-  context_.Say(*context_.location(),
-      "No matching construct for %s statement"_err_en_US, stmtType);
+  if (isExit) {
+    context_.Say(*context_.location(),
+        "No matching construct for EXIT statement"_err_en_US);
+  } else {
+    context_.Say(*context_.location(),
+        "No matching DO construct for CYCLE statement"_err_en_US);
```

Comment:
Can't you omit the first arg from these two calls to `Say()`.

Reply:
Yep.  I'll remove them.